A few Python API changes

paulromano opened the pull request:
This pull request changes how scores, nuclides, filters, and triggers are added to a tally. Namely, rather than relying on `add_*` methods, there are now scores, nuclides, filters, and triggers properties that expect lists (more precisely, mutable sequences). This allows you to do the following:
```Python
tally.scores = ['total', 'fission', 'absorption']
```
whereas before it would have required you to call add_score three times or do a @wbinventor -style
```Python
map(tally.add_score, ['total', 'fission', 'absorption'])
```
I should note that the `add_*` methods still exist and raise a DeprecationWarning.

Two other changes:
- Executor uses `universal_newlines=True`. If you've ever tried to execute OpenMC using Python 3 you'll see it's not pretty. That's now fixed.
- I've removed the warnings filter which forces our DeprecationWarnings to be displayed once. The problem is that this causes DeprecationWarnings to show up from all packages, which is probably something a user of our API doesn't necessary care for.

wbinventor commented:
Very good call on this @paulromano! I'll review this shortly. A few things that you might consider:

* Use the new properties in the Jupyter Notebook examples
* Use the new properties in the example input files
* Remove the deprecation warnings on `add_*(...)` methods? These methods can still be very useful (and cleaner looking than calling `append` on the properties themselves) for automated workflows in which the full list of filters, nuclides and scores are not known a priori and are instead added to a tally incrementally.

smharper commented:
I don't like the idea of having two equally supported methods for adding filters, etc.  It sounds like extra code and confusion for very little benefit.

paulromano commented:
I agree with @smharper. So does the Zen of Python:

    There should be one-- and preferably only one --obvious way to do it.
    Although that way may not be obvious at first unless you're Dutch.

wbinventor commented:
Is calling the `append` method on the underlying list "obvious?" Agreed
that we should eliminate the DoF to do the same things, but requiring all
filters, nuclides, scores to be assigned at once can make complicated input
generation and downstream data processing code very convoluted.
On Feb 23, 2016 11:29 AM, "Paul Romano" <notifications@github.com> wrote:

> I agree with @smharper <https://github.com/smharper>. So does the Zen of
> Python:
>
> There should be one-- and preferably only one --obvious way to do it.
> Although that way may not be obvious at first unless you're Dutch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/593#issuecomment-187774563>.
>

paulromano commented:
The class documentation says that those attributes are to be lists, so I would say yes, calling append to add another item is "obvious". And we are certainly not requiring that all scores, filters, etc. be assigned at once. To add one at a time, use the list as it is intended to be used.

paulromano commented:
@wbinventor Just updated the notebooks and the examples were already updated.

paulromano commented:
I also just added a `.. deprecated::` directive to the documentation of the add_* methods which instructs the user to use the properties directly. Hopefully that will make it more obvious.

wbinventor commented:
Cool, I'll adapt to the new and improved realities of the properties then.

paulromano commented:
@wbinventor Was there something else you wanted me to change before merging?

wbinventor commented:
Sorry, I haven't had an opportunity to review this at all yet since I was
away for a long weekend. I'll take a look at this tomorrow.

On Mon, Feb 29, 2016 at 11:23 PM, Paul Romano <notifications@github.com>
wrote:

> @wbinventor <https://github.com/wbinventor> Was there something else you
> wanted me to change before merging?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/593#issuecomment-190534156>.
>



-- 
Will Boyd
Nuclear Science & Engineering
Massachusetts Institute of Technology
Cell: 423.413.8469
wbinventor@gmail.com

wbinventor commented on docs/source/pythonapi/examples/tally-arithmetic.ipynb:
```diff
@@ -418,29 +418,25 @@
-    "tally.add_nuclide(u238)\n",
-    "tally.add_nuclide(u235)\n",
+    "tally.filters = [openmc.Filter(type='cell', bins=[fuel_cell.id]),\n",
+    "                 energy_filter]\n",
```
How about using `append` here on the continuation line as you do below? I think 1) it looks cleaner and 2) it illustrates that this can be done.

wbinventor commented on docs/source/pythonapi/examples/tally-arithmetic.ipynb:
```diff
@@ -418,29 +418,25 @@
-    "tally.add_filter(openmc.Filter(type='cell', bins=[fuel_cell.id, moderator_cell.id]))\n",
-    "tally.add_filter(energy_filter)\n",
-    "tally.add_score('flux')\n",
+    "tally.filters = [openmc.Filter(type='cell', bins=[fuel_cell.id, moderator_cell.id]),\n",
```
Ditto use `append` for the continuation line with `energy_filter`

wbinventor commented on docs/source/pythonapi/examples/tally-arithmetic.ipynb:
```diff
@@ -486,9 +482,9 @@
-    "fuel_therm_abs_rate.add_filter(openmc.Filter(type='energy', bins=[0., 0.625e-6]))\n",
-    "fuel_therm_abs_rate.add_filter(openmc.Filter(type='cell', bins=[fuel_cell.id]))\n",
+    "fuel_therm_abs_rate.scores = ['absorption']\n",
+    "fuel_therm_abs_rate.filters = [openmc.Filter(type='energy', bins=[0., 0.625e-6]),\n",
```
Up to you whether to use `append` here, my OCD self is less opinionated if the continuation line length is of the same order as the preceding line :-)

wbinventor commented on docs/source/pythonapi/examples/tally-arithmetic.ipynb:
```diff
@@ -520,12 +516,10 @@
-    "tally.add_nuclide(h1)\n",
-    "tally.add_nuclide(u238)\n",
+    "tally.filters = [openmc.Filter(type='cell', bins=[fuel_cell.id, moderator_cell.id]),\n",
+    "                 energy_filter]\n",
```
perhaps `append`

wbinventor commented on examples/python/basic/build-xml.py:
```diff
@@ -109,29 +109,20 @@
 first_tally = openmc.Tally(tally_id=1, name='first tally')
-first_tally.add_filter(cell_filter)
-scores = ['total', 'scatter', 'nu-scatter', \
+first_tally.filters = [cell_filter]
```
You've got me sold, I like the simplification now :-)

wbinventor commented on openmc/checkvalue.py:
```diff
@@ -41,9 +41,9 @@ def check_type(name, value, expected_type, expected_iter_type=None):
+    expected_type : type or Iterable of type
         type to check object against
-    expected_iter_type : type or None, optional
+    expected_iter_type : type or Iterable of type or None, optional
```
Nice, I've thought for some time now that such extensions would be useful, but hadn't run into a use case which really needed it yet. But I can see why this PR could make use of this, as well as future improvements to the API.

wbinventor commented on openmc/executor.py:
```diff
@@ -27,7 +27,8 @@ def _run_openmc(self, command, output):
                              cwd=self._working_directory,
-                             stdout=subprocess.PIPE)
+                             stdout=subprocess.PIPE,
+                             universal_newlines=True)
```
It's nice that the fix for Python 3 was so simple.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -510,27 +510,27 @@ def _create_tallies(self, scores, all_filters, keys, estimator):
         for score, key, filters in zip(scores, keys, all_filters):
             self.tallies[key] = openmc.Tally(name=self.name)
-            self.tallies[key].add_score(score)
+            self.tallies[key].scores.append(score)
```
Although I see why this works, I do believe that `score` will be a single `str`, and `filters` will be a list of filters, so the following should work:

```python
self.tallies[key].scores = [score]
self.tallies[key].filters = filters
```

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -510,27 +510,27 @@ def _create_tallies(self, scores, all_filters, keys, estimator):
                 trigger_clone = copy.deepcopy(self.tally_trigger)
-                trigger_clone.add_score(score)
-                self.tallies[key].add_trigger(trigger_clone)
+                trigger_clone.scores.append(score)
```
Ditto, perhaps:

```python
trigger_clone.scores = [scores]
```

though this isn't particularly important.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -510,27 +510,27 @@ def _create_tallies(self, scores, all_filters, keys, estimator):
             # Add all non-domain specific Filters (e.g., 'energy') to the Tally
             for add_filter in filters:
-                self.tallies[key].add_filter(add_filter)
+                self.tallies[key].filters.append(add_filter)
```
Ok, I can see that you were likely using `append` above to be consistent with this, so if you want to keep `append` everywhere in this method that is fine with me.

wbinventor commented on openmc/statepoint.py:
```diff
@@ -389,7 +389,7 @@ def tallies(self):
 
                     # Add Filter to the Tally
-                    tally.add_filter(new_filter)
+                    tally.filters.append(new_filter)
```
My issue with using `append` in something so fundamental as `StatePoint` is no error checking is performed (albeit everything *should* be fine without error checking, it has been useful for me during the early stages of development of the API to have some error checking here). What do you think about reworking this code to create a list of `Filter` and then assign that list to `tally.filters`?

paulromano replied:
This is a very good point. I had wrongly assumed that calling append would effectively trigger the property setter, but it seems that it doesn't. Thinking about it again now, it makes sense that it wouldn't. In theory, we could subclass list and override the append method, forcing it to typecheck its elements and then have the setter always convert into the list subclass, but that seems a little overkill. Hmm, let me think more about this.

wbinventor replied:
Looks like you wrote this while I was typing up my final comment below. I think our thought processes are converging here and will let you decide what's best.

wbinventor commented on openmc/statepoint.py:
```diff
@@ -398,7 +398,7 @@ def tallies(self):
                 for name in nuclide_names:
                     nuclide = openmc.Nuclide(name.decode().strip())
-                    tally.add_nuclide(nuclide)
+                    tally.nuclides.append(nuclide)
```
ditto, same comment about direct assignment to `tally.nuclides`

wbinventor commented on openmc/statepoint.py:
```diff
@@ -425,7 +425,7 @@ def tallies(self):
                     score = re.sub(pattern, '-' + moments[j].decode(), score)
 
-                    tally.add_score(score)
+                    tally.scores.append(score)
```
ditto, same comment about direct assignment to `tally.scores`

wbinventor commented on openmc/summary.py:
```diff
@@ -539,7 +539,7 @@ def _read_tallies(self):
                 pattern = r'-n$|-pn$|-yn$'
                 score = re.sub(pattern, '-' + moments[j].decode(), score)
-                tally.add_score(score)
+                tally.scores.append(score)
```
ditto, same comment about direct assignment to `tally.scores`

wbinventor commented on openmc/summary.py:
```diff
@@ -560,7 +560,7 @@ def _read_tallies(self):
 
                 # Add Filter to the Tally
-                tally.add_filter(new_filter)
+                tally.filters.append(new_filter)
```
ditto, same comment about direct assignment to `tally.filters`

wbinventor commented on openmc/tallies.py:
```diff
@@ -433,13 +443,11 @@ def add_trigger(self, trigger):
-
-        if trigger not in self.triggers:
-            self.triggers.append(trigger)
+        warnings.warn("Tally.add_trigger(...) has been deprecated and may be "
```
Not a big deal, but all strings are defined with single quotes elsewhere in the API, may want to do the same here for consistency.

wbinventor commented on openmc/tallies.py:
```diff
@@ -475,23 +536,19 @@ def add_filter(self, new_filter):
-            raise ValueError(msg)
-
-        self._filters.append(new_filter)
+        warnings.warn("Tally.add_filter(...) has been deprecated and may be "
```
ditto single quotes

wbinventor commented on openmc/tallies.py:
```diff
@@ -504,24 +561,19 @@ def add_nuclide(self, nuclide):
-            raise ValueError(msg)
-
-        self._nuclides.append(nuclide)
+        warnings.warn("Tally.add_nuclide(...) has been deprecated and may be "
```
ditto single quotes

wbinventor commented on openmc/tallies.py:
```diff
@@ -533,24 +585,11 @@ def add_score(self, score):
-        # CrossScores and AggrgateScore
-        else:
-            self._scores.append(score)
+        warnings.warn("Tally.add_score(...) has been deprecated and may be "
```
ditto single quotes

wbinventor commented on openmc/tallies.py:
```diff
@@ -2210,12 +2249,9 @@ def __add__(self, other):
-                new_tally.add_nuclide(nuclide)
-            for score in self.scores:
-                new_tally.add_score(score)
+            new_tally.filters = self.filters
```
These lists need to be deep copied here. Otherwise if the user (or tally arithmetic) changes the scores in `self` or `new_tally` it will erroneously change them in the other.

wbinventor commented on openmc/tallies.py:
```diff
@@ -2284,12 +2320,9 @@ def __sub__(self, other):
-                new_tally.add_nuclide(nuclide)
-            for score in self.scores:
-                new_tally.add_score(score)
+            new_tally.filters = self.filters
```
ditto, deep copy the lists of filters, scores and nuclides

wbinventor commented on openmc/tallies.py:
```diff
@@ -2359,12 +2392,9 @@ def __mul__(self, other):
-                new_tally.add_nuclide(nuclide)
-            for score in self.scores:
-                new_tally.add_score(score)
+            new_tally.filters = self.filters
```
ditto, deep copy the lists of filters, scores and nuclides

wbinventor commented on openmc/tallies.py:
```diff
@@ -2434,12 +2464,9 @@ def __truediv__(self, other):
-                new_tally.add_nuclide(nuclide)
-            for score in self.scores:
-                new_tally.add_score(score)
+            new_tally.filters = self.filters
```
ditto, deep copy the lists of filters, scores and nuclides

wbinventor commented on openmc/tallies.py:
```diff
@@ -2513,12 +2540,9 @@ def __pow__(self, power):
-                new_tally.add_nuclide(nuclide)
-            for score in self.scores:
-                new_tally.add_score(score)
+            new_tally.filters = self.filters
```
ditto, deep copy the lists of filters, scores and nuclides

wbinventor commented on openmc/trigger.py:
```diff
@@ -107,16 +117,11 @@ def add_score(self, score):
-            return
-        else:
-            self._scores.append(score)
+        warnings.warn("Trigger.add_score(...) has been deprecated and may be "
```
ditto single quotes

wbinventor commented:
@paulromano I've finished my review of this PR and overall I like the changes made here a lot and we should be able to merge this is in soon. Based on my review, however, I am still torn about removing the `add_*` methods entirely in v0.8. Based on my count, the API itself now uses `append` without error checking scores, nuclides and filters in 43 different locations (not counting examples or code comments). The extensive use of `append` in the tally arithmetic code most troubles since it was not a simple exercise to implement this for general use, nor is it simple to debug (as I'm sure @samuelshaner can concur). 

By moving to simple `append` we now put the onus on the developer to "get it right" with their code in the API or else it may be more difficult to to debug the API. While using `append` vs. `add_*` methods in things like tally arithmetic is only one small door through which faulty implementation bugs may enter, it is indeed a plausible one and one which I faced when implementing that code in the first place. For this reason, I would still vote to keep the `add_*` methods around (perhaps make them "private" as `_add_*` methods), but am certainly fine with moving on if others disagree.

paulromano commented:
@wbinventor I believe I've addressed all your comments where appropriate. My solution for the `append` problem was to create a new `CheckedList` class which overrides `append` as well as `insert` (in case a user changes the list that way too). Now direct assignment and `append` are "safe".

I also noticed that a recent change you made in mgxs created a dependency on pandas >= 0.17 (use of `DataFrame.sort_values(...)`). No problem, but just an FYI. If someone other than you is using this stuff, they are probably pretty cutting edge anyway.

wbinventor commented:
@paulromano thanks for the updates, your solution with `CheckedList` is the perfect compromise between simplicity and safety. I'll merge this in once Travis finishes its run.

As for Pandas 0.17.1, the move to use `sort_values(...)` was to eliminate some deprecation warnings. Pandas is rapidly developed and 0.17 is the default version for Anaconda (and pip I think), so I don't expect this dependency to be an issue for anyone within a matter of a few months. But I apologize for not realizing this and including it in my earlier PR.
